### PR TITLE
DYNAMIC_HYBRID Find mode

### PR DESCRIPTION
**Pull request by atamazov**
This implements the DYNAMIC_HYBRID Find mode, see https://github.com/ROCmSoftwarePlatform/MIOpen/issues/398#issuecomment-687210054.

- Added DYNAMIC_HYBRID mode 
- Added/initialized the `use_dynamic_solutions_only` attribute in the convolution context
- Respective Solvers marked as as Dynamic:
  - ConvAsmImplicitGemmV4R1DynamicFwd
  - ConvAsmImplicitGemmV4R1DynamicFwd_1x1
  - ConvBinWinograd3x3U
  - ConvBinWinogradRxS
  - ConvBinWinogradRxSf3x2
  - ConvBinWinogradRxSf2x3
  - ConvMPBidirectWinograd
  - ConvWinograd3x3MultipassWrW
- In case of find-db miss, Find and GWSS use only dynamic solutions
  - GEMM is not affected.
  - :warning: FFT algorithm is not affected.
- Documentation updated

## :cyclone: Local performance & correctness testing

- Radeon VII, ROCm 3.5.
- 93 known FP32 + 93 known FP16 convolution configs
- Normal Find mode + Immediate mode
- OCL backend (FP32 configs only) + HIP backend
- Develop baseline (c7428da0c) vs this PR (d2f9ecb38)

No regressions.

## :cyclone: Evaluation of performance impact

- Radeon VII, ROCm 3.5.
- 93 unknown FP32 convolution configs
- HIP backend
- This PR (d2f9ecb38)

Find Mode | Wall time acceleration | GPU_perf_drop | Linux real time, sec | Linux real time acceleration
-- | -- | -- | -- | --
NORMAL | 1.00 | 0% | 926 | 1
FAST | 26.18 | 52% | 278 | 3.33
HYBRID | 1.00 | 0% | 925 | 1
FAST_HYBRID | 1.57 | 1% | 684 | 1.35
DYNAMIC_HYBRID | 5.13 | 21% | 333 | 2.78

DYNAMIC_HYBRID is similar to FAST, that is what we expected.

Note that numbers obtained with BUILD_DEV=On and binary cache disabled. This is far from what happens when the real framework is running. However, this should be consistent with a situation where kernels are always compiled, i.e. in real life the trends should be similar.

### CODE REVIEW

**Comment by atamazov** [comment]
"Evaluation of performance impact" is ready.

**Comment by atamazov** [comment]
GWSS should also observe `use_dynamic_only`.

**Review by JehandadKhan — CHANGES_REQUESTED** [review]
Please add the following solvers as well (if they are dynamic)

```c++
ConvAsmImplicitGemmV4R1DynamicBwd
ConvAsmImplicitGemmV4R1DynamicWrw
```

**Comment by atamazov** [comment]
Ready for another review round.

**Comment by atamazov** [comment]
"Evaluation of performance impact" updated.

**Comment by atamazov** [comment]
"Local performance & correctness testing" updated.

File: src/include/miopen/find_solution.hpp [hunk 1]
```diff
@@ -145,7 +145,11 @@ struct SolverContainer
                 if(find_only.IsValid() && find_only != Id{SolverDbId(solver)})
                 { // Do nothing (and keep silence for the sake of Tuna), just skip.
                 }
-                else if(solver.IsApplicable(search_params))
+                else if(!solver.IsApplicable(search_params))
+                    MIOPEN_LOG_I2(SolverDbId(solver) << ": Not applicable");
+                else if(search_params.use_dynamic_solutions_only && !solver.IsDynamic())
+                    MIOPEN_LOG_I(SolverDbId(solver) << ": Skipped (non-dynamic).");
```
[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/find_solution.hpp]: Done

**Comment by JehandadKhan** [comment]
@atamazov Do you intend to open a separate PR for the heuristic code place holder ? We can add the called function now, but it would just return. Knowing where the function is would make @zjing14's job much easier.

**Comment by atamazov** [comment]
@JehandadKhan Yes, there will be separate PR for WTI based heuristic (in progress, ETA this week).

Let's put the MLP classifier into separate module (.cpp with public interface in .hpp). I will integrate it on top of that PR.

**Review by JehandadKhan — APPROVED** [review]

**Comment by JehandadKhan** [comment]
@daniellowell  Lets merge this in.

**Pull request merged**